presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -1403,22 +1406,24 @@ private void rollbackShared()
             switch (declaredIntentionToWrite.getMode()) {
                 case STAGE_AND_MOVE_TO_TARGET_DIRECTORY:
                 case DIRECT_TO_TARGET_NEW_DIRECTORY: {
-                    // Note: there is no need to cleanup the target directory as it will only be written
-                    // to during the commit call and the commit call cleans up after failures.
-                    Path rootPath = declaredIntentionToWrite.getRootPath();
-
-                    // In the case of DIRECT_TO_TARGET_NEW_DIRECTORY, if the directory is not guaranteed to be unique
-                    // for the query, it is possible that another query or compute engine may see the directory, wrote
-                    // data to it, and exported it through metastore. Therefore it may be argued that cleanup of staging
-                    // directories must be carried out conservatively. To be safe, we only delete files that start with
-                    // the unique prefix for queries in this transaction.
-
-                    recursiveDeleteFilesAndLog(
-                            declaredIntentionToWrite.getContext(),
-                            rootPath,
-                            ImmutableList.of(declaredIntentionToWrite.getFilePrefix()),
-                            true,
-                            format("staging/target_new directory rollback for table %s", declaredIntentionToWrite.getSchemaTableName()));
+                    if (!skipTargetCleanupOnFailure || declaredIntentionToWrite.getMode() != DIRECT_TO_TARGET_NEW_DIRECTORY) {
```

Comment:
It's possible for a connector to sometimes use DIRECT_TO_TARGET_NEW_DIRECTORY and sometimes not.  e.g. HiveLocationService uses a new directory for a new table, and for an existing table either does a staging table or direct to existing directory depending on session properties and file system.  So, we don't want to fail just because the current transaction doesn't take advantage of the property.

Reply:
I was rather thinking about "fail fast". So basically check on the startup that if not `DIRECT_TO_TARGET_NEW_DIRECTORY` and `skipTargetCleanupOnFailure` -> fail on startup

Reply:
Totally missed that. Yeah, that makes sense.

Reply:
I'm wondering if the write mode is selected by the location service, should the rollback behaviour be set by the location service as well?

CC: @wenleix @highker ?

Reply:
Since the rest of the rollback is done independently of the location service, I think defining this aspect of clean up in the location service would split the rollback responsibility into two places and be more confusing.  The rollback mode is related to the location service, but there isn't really a dependency.

Reply:
Here are my thoughts. Location service may communicate with the remote services, that may decided if the staging data has to be removed by Presto. It looks like we are splitting configuration into 2 different places. But i don't have a strong opinion on that.